stl/inc/algorithm

```diff
@@ -2364,6 +2366,149 @@ namespace ranges {
+                return false;
+            } else if constexpr (_Sized_or_infinite_range<_Rng1> && sized_range<_Rng2>) {
+                const range_difference_t<_Rng2> _Count2 = _RANGES distance(_Range2);
+                if constexpr (sized_range<_Rng1>) {
+                    if (_Count2 > _RANGES distance(_Range1)) {
+                        return false;
+                    }
+                }
+
+                return _RANGES _Equal_count(_Ubegin(_Range1), _Ubegin(_Range2),
+                    static_cast<range_difference_t<_Rng1>>(_Count2), _Pass_fn(_Pred), _Pass_fn(_Proj1),
+                    _Pass_fn(_Proj2));
+            } else {
+                const auto _ULast2 = _Uend(_Range2);
+                const auto _Result = _RANGES _Mismatch_4(_Ubegin(_Range1), _Uend(_Range1), _Ubegin(_Range2), _ULast2,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+                return _Result.in2 == _ULast2;
+            }
+        }
+    };
+
+    inline constexpr _Starts_with_fn starts_with{_Not_quite_object::_Construct_tag{}};
+
+    // VARIABLE ranges::ends_with
+    class _Ends_with_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, input_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pr = ranges::equal_to, class _Pj1 = identity, class _Pj2 = identity>
+            requires (forward_iterator<_It1> || sized_sentinel_for<_Se1, _It1>)
+                  && (forward_iterator<_It2> || sized_sentinel_for<_Se2, _It2>)
+                  && indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr bool operator()(_It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred = {},
+            _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            // clang-format on
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+
+            if constexpr (same_as<_Se1, unreachable_sentinel_t> || same_as<_Se2, unreachable_sentinel_t>) {
+                return false;
+            } else {
+                iter_difference_t<_It1> _Count1;
+                if constexpr (sized_sentinel_for<_Se1, _It1>) {
+                    _Count1 = _Last1 - _First1;
+                } else {
+                    _Count1 = _RANGES distance(_First1, _Last1);
+                }
+
+                iter_difference_t<_It2> _Count2;
+                if constexpr (sized_sentinel_for<_Se2, _It2>) {
+                    _Count2 = _Last2 - _First2;
+                } else {
+                    _Count2 = _RANGES distance(_First2, _Last2);
+                }
+
+                if (_Count2 > _Count1) {
+                    return false;
+                }
+
+                auto _UFirst1 = _Get_unwrapped(_STD move(_First1));
+                _RANGES advance(_UFirst1, (_Count1 - _Count2));

```

Comment:
This could be implicitly narrowing `_Count1 - _Count2` to `iter_difference_t<_It1>`, which will produce compiler warnings.  We should `static_cast` to avoid that, perhaps in a separate step (`_Count1 = static_cast<iter_difference_t<_It1>>(_Count1 - _Count2)`) to keep things simple where we can.

Reply:
Hm, I do not really see a benefit in the additional assignment. Shout if your really want it